### PR TITLE
Implement happy eyeballs (RFC 8305)

### LINKED ISSUE #4451: Support Happy Eyeballs from Python 3.8
Starting from Python 3.8 asyncio supports https://tools.ietf.org/html/rfc6555

It allows working better in dual-host configurations.
It would be nice to support it by aiohttp if the feature is available by Python version.
Perhaps passing `happy_eyeballs_delay` and `interleave` to `loop.create_connection()` is enough.

A volunteer is welcome!

P.S.
Ooops, parameters are even not documented by asyncio docs, need to fix it.

### PR DESCRIPTION
**Pull request by bdraco**
## What do these changes do?

Implements Happy Eyeballs support (https://datatracker.ietf.org/doc/html/rfc8305)

## Are there changes in behavior for the user?

Dual stack connections where IPv6 is not working should now be successful.

## Related issue number

fixes #4451

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

--

TODO

- [x] docs (params `happy_eyeballs_delay` and `interleave` function the same as cpython but we should point that out, and how to disable `happy_eyeballs_delay` by setting it to `None`)
- [x] more production and functional testing
- [x] tests for ipv4 fallback
- [x] tests for interleave
- [x] tests for family
- [x] timeline

This PR is a first step as I need to make sure this library works for the other projects as well. Its likely to sit for a bit while I test the library with all the projects that need it so there aren't any need to do breaking changes before all the other projects start consuming it:
- https://github.com/esphome/aioesphomeapi/pull/789
- https://github.com/Jc2k/aiohomekit/pull/349

--

### CODE REVIEW

File: aiohttp/connector.py
Comment (bdraco) [inline aiohttp/connector.py]: Should probably be a helper

File: aiohttp/connector.py
Comment (bdraco) [inline aiohttp/connector.py]: This isn't great, would be better to make a helper here

**Comment by bdraco** [comment]
functional tests seem to be ok

Anything that does this will fail since `aiohappyeyeballs.start_connection` isn't mocked
```
        self.loop.create_connection = make_mocked_coro((mock.Mock(), mock.Mock()))
```

File: aiohttp/connector.py
Comment (bdraco) [inline aiohttp/connector.py]: This should probably raise instead of popping the first one since it means we can't find the bad peer for some reason.  Likely a problem comparing the addresses since the str may be formatted different. 

Need to add coverage for this

File: requirements/constraints.txt
Comment (bdraco) [inline requirements/constraints.txt]: TODO: revert header changes manually

File: requirements/dev.txt
Comment (bdraco) [inline requirements/dev.txt]: TODO: revert header changes manually

File: requirements/test.txt
Comment (bdraco) [inline requirements/test.txt]: TODO: revert header changes manually

File: tests/conftest.py
Comment (bdraco) [inline tests/conftest.py]:
```suggestion
```

File: aiohttp/connector.py
Comment (bdraco) [inline aiohttp/connector.py]: This should take the whole peer since it could mean we have different ports even if it's unlikely

Comment (bdraco) [reply aiohttp/connector.py]: https://bugs.python.org/issue42620

File: requirements/test.txt
Comment (Dreamsorcerer) [reply requirements/test.txt]: I wonder if it's reasonable to add `--no-header`, so this causes less issues in future...

File: setup.cfg
Comment (Dreamsorcerer) [inline setup.cfg]: Should we maybe put this under speedups and make it optional?

Comment (bdraco) [reply setup.cfg]: I thought about this from the perspective that IPv6 support should be a base feature, and without Happy Eyeballs, IPV6 tends to get turned off because it cannot establish a connection before the timeout with many network setups.

At least from my experience with Home Assistant, we have IPv6 turned off in many places by setting the address family to AF_INET only because the lack of Happy Eyeballs makes using IPv6 untenable.

From my perspective, it is much more valuable if IPv6 works out of the box with the developer not having to tinker to get things working.

Comment (Dreamsorcerer) [reply setup.cfg]: OK, seems reasonable then. Hopefully we can get it into cpython in the next couple of releases and then drop the dependency again.

**Review by Dreamsorcerer — APPROVED** [review]

**Comment by bdraco** [comment]
After a brief detour to fix some mdns bugs in zeroconf, I've got the other libraries that need this working as expected.

I need to do more functional testing and update the connector tests

File: tests/test_connector.py
Comment (bdraco) [inline tests/test_connector.py:601]: We need a similar test to make sure it falls back to the ipv4 host but should be quite light as to not reimplement testing the library

File: aiohttp/connector.py
Comment (bdraco) [inline aiohttp/connector.py]: This isn't aiohttp specific so it could be moved to aiohappyeyeballs

File: aiohttp/connector.py
Comment (bdraco) [inline aiohttp/connector.py]: We could store the addresses in the cache in python's `getaddrinfo` format to avoid having to convert.  It would be more efficient and more portable.... but its a bit less readable.  Something to consider for the future though

File: aiohttp/connector.py
Comment (bdraco) [inline aiohttp/connector.py]: `is_ipv6_address` is likely overkill here as we don't need to validate it here, only determine the family

File: aiohttp/connector.py
Comment (bdraco) [reply aiohttp/connector.py]: This one is a bit of a special case though since ipv6 is always a 4 arg tuple, and this only takes a 2 arg tuple

Comment (bdraco) [reply aiohttp/connector.py]: Need to see how cpython handles this

Comment (bdraco) [reply aiohttp/connector.py]: https://github.com/python/cpython/blob/9cdf05bc28c5cd8b000b9541a907028819b3d63e/Lib/asyncio/base_events.py#L1472

Comment (bdraco) [reply aiohttp/connector.py]: cpython handles it the exact same way, it adds 0 0
https://github.com/python/cpython/blob/9cdf05bc28c5cd8b000b9541a907028819b3d63e/Lib/asyncio/base_events.py#L99C51-L99C72

So this is fine

File: aiohttp/connector.py
Comment (bdraco) [reply aiohttp/connector.py]: We don't need validation, only to get the right family

9e36ed1d830d5d6f02cc79cfd4c2a1777f22fba5

File: aiohttp/connector.py
Comment (bdraco) [reply aiohttp/connector.py]: Moved to the lib since its useful elsewhere and not aiohttp specific

**Comment by bdraco** [comment]
At this point everything is working how I expect it to with functional testing.

The `aiohappyeyeballs` should be bug for bug compatible with cpython (hopefully there aren't any obvious ones) since its almost 100% the same code as `create_connection` with only added typing and unneeded code removed.

I've asked some of the Home Assistant devs to take a look at all the PRs to see if I missed anything that might require a breaking change in the `aiohappyeyeballs` library later

File: aiohttp/connector.py
Comment (bdraco) [inline aiohttp/connector.py]: This value is completely ignored since it's only used as a result of dns resolution which is not done here.  We should remove it as it's useless here

Comment (bdraco) [inline aiohttp/connector.py]:
```suggestion
                    "",
```

File: tests/conftest.py
Comment (Dreamsorcerer) [inline tests/conftest.py]: If possible, it would be preferred to use `autospec=True, spec_set=True`, to minimise the possibility of misusing the mock.

File: tests/test_connector.py
Comment (Dreamsorcerer) [inline tests/test_connector.py]: Similar with mocks like this. If we have a class to base it off, then we can hopefully do `create_autospec(Socket, spec_set=True, instance=True)`

Comment (bdraco) [reply tests/test_connector.py]: 349203019c80ce4af4e3514eff3e8976c4e485a5

File: tests/conftest.py
Comment (bdraco) [reply tests/conftest.py]: 349203019c80ce4af4e3514eff3e8976c4e485a5

**Comment by bdraco** [comment]
I'm still doing testing with this, and I think its ready to move to wider production testing.

I'm going to spend a few cycles on `test_unsupported_upgrade` in https://github.com/aio-libs/aiohttp/pull/7960 but its a bit difficult without a windows box

**Comment by bdraco** [comment]
So far so good on production 3.9+backported.

Doing performance testing now

**Comment by bdraco** [comment]
performance looks good. All the time is nearly `loop.sock_connect` so should have no significant change

**Comment by bdraco** [comment]
Performance and production testing looks good.

I've got reviews on the other use cases so I think this is good to go for a final review.

I want to sleep on it and look it all over in the morning end to end again though before I mark it ready

**Comment by bdraco** [comment]
Still testing out great this morning

File: requirements/test.txt
Comment (bdraco) [reply requirements/test.txt]: That seems reasonable to do in another PR 👍

**Comment by bdraco** [comment]
I think this is good to merge at this point, but I'm slow rolling it because ESPHome is currently in beta which also uses the `aiohappyeyeballs` lib so no rush in merging as I'm happy to wait for the beta feedback in case anything has to change with the lib, but it now seems unlikely at this point.

**Comment by bdraco** [comment]
ESPHome is releasing today with the new lib, HA is doing beta next week with it

**Comment by bdraco** [comment]
This change in esphome release seems to be going swimmingly.  Next week we get beta feedback on it in aiohomekit from Home Assistant

**Comment by bdraco** [comment]
The changes in aiohomekit seem to be going well in the HA beta.

**Comment by bdraco** [comment]
Ha beta has been without issues in this area.  The full release happens tomorrow.  As long as everything goes smoothly, I'll merge this shortly after

**Comment by bdraco** [comment]
HA release shipped today.  No issues so far.  All looking good

**Comment by bdraco** [comment]
Everything seems solid with the HA release related to this. 

That's the last hurdle.

**Pull request merged**
